<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 

In [175]:
connection = psycopg2.connect(
    dbname = '',
    user = '',
    host = '',
    password = '',
    port = 
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
query_3_1 = f'''
    SELECT
    COUNT(id)
    FROM vacancies
'''

In [5]:
df = pd.read_sql_query(query_3_1, connection)
print(f'В базе {df.loc[0][0]} вакансий')

В базе 49197 вакансий


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2718051745.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
query_3_2 = f'''
SELECT
COUNT(id)
FROM employers
'''

In [7]:
df = pd.read_sql_query(query_3_2, connection)
print(f'В базе {df.loc[0][0]} работодатель')

В базе 23501 работодатель


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\1095439466.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
query_3_3 = f'''
SELECT
COUNT(id)
FROM areas
'''

In [9]:
df = pd.read_sql_query(query_3_3, connection)
print(f'В базе {df.loc[0][0]} региона')

В базе 1362 региона


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\827654143.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query_3_4 = f'''
SELECT
COUNT(id)
FROM industries 
'''

In [11]:
df = pd.read_sql_query(query_3_4, connection)
print(f'В базе {df.loc[0][0]} сферы деятельности')

В базе 294 сферы деятельности


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\1311782874.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


***

Представлено 5 баз данных. Первая с данными о вакансиях, включающая 49197 позиции. Вторая содержит информацию о работодателях (23501 компании). Третья база включает в себя данные о регионах (1362 объекта). Четвертая содержит информацию об отраслях (294 отрасли / сферы деятельности), а при помощи пятой базы данных определяется отраслевая принадлежность работодателей.
Можно обратить внимание на то, что на одного работодателя приходится в среднем 2,1 вакансии, что говорит либо о том, что выборка состоит из достаточно маленьких компаний, либо о том, что по многим работодателям мы не найдем ни одной вакансии.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
n = 5
query_4_1 = f'''
SELECT
a.name as "Регион",
COUNT(*) as cnt
FROM vacancies v
JOIN areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC
LIMIT {n}
'''

In [13]:
df = pd.read_sql_query(query_4_1, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2004364207.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


Регион   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
query_4_2 = f'''
SELECT
COUNT(*)
FROM vacancies
WHERE salary_from IS NOT NULL or salary_to IS NOT NULL
'''

In [15]:
df = pd.read_sql_query(query_4_2, connection)
print(f'У {df.loc[0][0]} вакансий указана либо верхняя, либо нижняя граница зарплаты')

У 24073 вакансий указана либо верхняя, либо нижняя граница зарплаты


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\1413837217.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
query_4_3 = f'''
SELECT
AVG(salary_from) as salary_from_avg,
AVG(salary_to) as salary_to_avg
FROM vacancies
'''

In [17]:
df = pd.read_sql_query(query_4_3, connection)
print(f'Среднее значение для нижней границы зарплаты {int(round(df.loc[0][0], 0))} рублей, среднее значение для верхней границы зарплаты {int(round(df.loc[0][1], 0))} рублей')

Среднее значение для нижней границы зарплаты 71065 рублей, среднее значение для верхней границы зарплаты 110537 рублей


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\3659867109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
query_4_4 = f'''
SELECT
schedule,
employment,
COUNT(*)
FROM vacancies
GROUP BY schedule, employment
ORDER BY 3 DESC
OFFSET 1
LIMIT 1
'''

In [19]:
df = pd.read_sql_query(query_4_4, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\4268832789.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule        employment  count
0  Удаленная работа  Полная занятость   7802

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
query_4_5 = f'''
SELECT
experience,
COUNT(*)
FROM vacancies
GROUP BY experience
ORDER BY 2
'''

In [21]:
df = pd.read_sql_query(query_4_5, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\3611053821.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

Ожидаемо большинство вакансий имеют локация в Москве и Санкт-Петербурге, также представлен Новосибирск. При этом в число распространенных городов попали Минск и Алматы.
Важно обратить внимание, что около половины (51%) вакансий не имеют никакого ориентира по зарплате. При этом средняя нижняя граница равна 71065 рублей, а средняя верхняя граница - 110537 рублей.
Работодатели преимущественно ищут сотрудников на полную занятость, но часто допускают удаленную работу (вторая по популярности опция). При этом в выборке присутствует значительная концентрация на сотрудниках с малым опытом (самая популярная группа с опытом от 1 до 3 лет, вторая по популярности - вообще без опыта; на обе эти группы приходится 83% вакансий).

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
n1 = 1
n2 = 5
query_5_1 = f'''(
SELECT
e.name,
COUNT(*)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
GROUP BY e.name
ORDER BY 2 desc
OFFSET {n1 - 1}
LIMIT 1)
    
UNION ALL
    
(SELECT
e.name,
COUNT(*)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
GROUP BY e.name
ORDER BY 2 desc
OFFSET {n2 - 1}
LIMIT 1)   
'''

In [23]:
df = pd.read_sql_query(query_5_1, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2915140023.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
query_5_2 = f'''
SELECT
a.name,
COUNT(DISTINCT v.id),
COUNT(DISTINCT e.id)
FROM areas a
LEFT JOIN employers e ON e.area = a.id
LEFT JOIN vacancies v ON v.area_id = a.id
WHERE v.id is NULL
GROUP BY a.name
ORDER BY 3 DESC
LIMIT 1
'''

In [25]:
df = pd.read_sql_query(query_5_2, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2644321718.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  count  count
0  Россия      0    410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
query_5_3 = f'''
SELECT
e.name,
COUNT(distinct v.area_id)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
GROUP BY e.name
ORDER BY 2 DESC
LIMIT 1
'''

In [27]:
df = pd.read_sql_query(query_5_3, connection)
print(f'Наибольшее количество регионов, в которых размещены вакансии, у {df.loc[0][0]}. Количество регионов составляет {df.loc[0][1]}.')

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\4277563636.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Наибольшее количество регионов, в которых размещены вакансии, у Яндекс. Количество регионов составляет 181.


4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
query_5_4 = f'''
SELECT
COUNT(*)
FROM employers e
LEFT JOIN employers_industries ei ON ei.employer_id = e.id
WHERE ei.industry_id is NULL
'''

In [29]:
df = pd.read_sql_query(query_5_4, connection)
print(f'У {df.loc[0][0]} работодателей не указана сфера детельности.')

У 8419 работодателей не указана сфера детельности.


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\3742368544.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
query_5_5 = f'''
SELECT
e.name,
COUNT(ei.industry_id)
FROM employers e
JOIN employers_industries ei ON ei.employer_id = e.id
GROUP BY e.name
HAVING COUNT(ei.industry_id) = 4
ORDER BY 1
OFFSET 2
LIMIT 1
'''

In [31]:
df = pd.read_sql_query(query_5_5, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\1337360513.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
query_5_6 = f'''
SELECT
DISTINCT i.name,
COUNT (*)
FROM employers_industries ei
JOIN industries i ON ei.industry_id = i.id
GROUP BY i.name
HAVING i.name = 'Разработка программного обеспечения'
ORDER BY 2 DESC
'''

In [33]:
df = pd.read_sql_query(query_5_6, connection)
print(f'У {df.loc[0][1]} работодателей в качестве сфера детельности указана Разработка прогнрамного обеспечения.')

У 3553 работодателей в качестве сфера детельности указана Разработка прогнрамного обеспечения.


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\302879986.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# выгрузка списка городов-миллионников со страницы в Википелия
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
cities = []
table_cities = pd.read_html(url)[1]
for i in table_cities['Город']:
    if '[' in i: # Проверка на наличие сносок
        i = i[:-3]
    cities.append(i)
cities = tuple(cities)

In [35]:
query_5_7 = f'''(
SELECT
a.name,
COUNT (*)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
JOIN areas a ON a.id = v.area_id
WHERE e.name = 'Яндекс' AND a.name in {cities}
GROUP BY a.name)

UNION ALL

SELECT
'Total',
COUNT (*)
FROM vacancies v
JOIN employers e ON e.id = v.employer_id
JOIN areas a ON a.id = v.area_id
WHERE e.name = 'Яндекс' AND a.name in {cities}

ORDER BY 2
'''


In [36]:
df = pd.read_sql_query(query_5_7, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\3290006367.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

Основное число вакансий приходится на Яндексом, однако его доля в общей выборке невелика - около 4%. При этом вакансии Яндекса достаточно равномерно распределены по городам (в т.ч. по городам-миллионникам), кроме того вакансии этой компании представленный в наибольшем количестве населенных пунктов. У многих компаний (36%) вообще не указана сфера деятельности. Среди основных сфер деятельности указаны - разработка программного обеспечения и системная интеграция, что в целом соответствует общей специфике рассматриваемой выборки.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
query_6_1 = f'''
SELECT
COUNT (*)
FROM vacancies
WHERE LOWER(name) LIKE '%data%' or LOWER(name) LIKE '%данн%'
'''


In [38]:
df = pd.read_sql_query(query_6_1, connection)
print(f'{df.loc[0][0]} вакансия имеют отношение к данным, т.е. в названии содержатся DATA или ДАНН')

1771 вакансия имеют отношение к данным, т.е. в названии содержатся DATA или ДАНН


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\1453647653.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [105]:
query_6_2 =  f'''
SELECT
COUNT (*)
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')
AND (LOWER(name) LIKE '%junior%' OR experience = 'Нет опыта' OR employment = 'Стажировка')
'''

In [106]:
df = pd.read_sql_query(query_6_2, connection)
print(f'В базе {df.loc[0][0]} подходящая для начинающего дата-сайентиста вакансия')

В базе 51 подходящая для начинающего дата-сайентиста вакансия


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2215426256.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
query_6_3 =  f'''
SELECT
COUNT (*)
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')
AND (key_skills ILIKE '%sql%' OR key_skills ILIKE '%postgres%')
'''

In [42]:
df = pd.read_sql_query(query_6_3, connection)
print(f'У {df.loc[0][0]} вакансии для DC в качестве ключевого навыка указаны SQL или postgres')

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2900919947.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


У 201 вакансии для DC в качестве ключевого навыка указаны SQL или postgres


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
query_6_4 =  f'''
SELECT
COUNT (*)
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')
AND (key_skills ILIKE '%python%')
'''

In [44]:
df = pd.read_sql_query(query_6_4, connection)
print(f'У {df.loc[0][0]} вакансии для DC в качестве ключевого навыка указан Python')

У 351 вакансии для DC в качестве ключевого навыка указан Python


C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\3696570521.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
query_6_5 =  f'''
SELECT
ROUND(AVG(LENGTH(key_skills) - LENGTH(REPLACE(key_skills, '\t', '')) + 1), 2)
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')
'''

In [46]:
df = pd.read_sql_query(query_6_5, connection)
print(f'{df.loc[0][0]} ключевых навыка в среднем указывают в вакансиях для DS')

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2777671892.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


6.41 ключевых навыка в среднем указывают в вакансиях для DS


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [178]:
query_6_6 =  f'''
SELECT
experience,
ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0) AS avg_slr
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')
GROUP BY experience
'''

In [179]:
df = pd.read_sql_query(query_6_6, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\4002724519.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience   avg_slr
0         Более 6 лет       NaN
1           Нет опыта   74643.0
2  От 1 года до 3 лет  139675.0
3       От 3 до 6 лет  243115.0

***

Из общей выборки 1771 вакансия (или 3,6%) имеет отношение к данным. Для дата-сайентистов начального уровня доступно 51 вакансия или 2,9% от всех вакансий, связанных с данными. 201 вакансия (11% от всех вакансий, связанных сданными) требуют навыков работы с SQL или postgres, владения Python требуется в 20% вакансий (351 вакансия). При этом от дата-сайентиста работодатели ожидают наличие в среднем 6,4 ключевых навыка. На начальном этапе своей карьеры дата-сайентист может рассчитывать на зарплату около 75 тыс. рублей, после года опыта показатель может увеличиться почти в два раза - до 140 тыс. рублей, тогда как после трех лет опыта зарплаты может приблизиться к 250 тыс. рублей.

# Общий вывод по проекту

Перед общим выводом по проекту рассмотрим еще несколько запросов, которые помогут лучше понимать особенности выборки.

1. Рассмотрим зависимость средней зарплаты от типа рабочего графика для дата-сайентистов.

In [141]:
query_7_1 =  f'''
SELECT
schedule,
ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0)::text avg_slr,
COUNT (id) cnt
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')
AND COALESCE((salary_from + salary_to) / 2, salary_from, salary_to) IS NOT NULL
GROUP BY schedule

UNION ALL

SELECT
'Итого по всем вакансиям DS с указанной ЗП',
ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to)), 0)::text,
COUNT (id)
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')
AND COALESCE((salary_from + salary_to) / 2, salary_from, salary_to) IS NOT NULL

UNION ALL

SELECT
'Итого по всем вакансиям DS',
' ',
COUNT (id)
FROM vacancies
WHERE (name ILIKE '%data%scientist%' OR name ILIKE '%data%science%' OR name ILIKE '%исследователь%данных%' OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
OR name ILIKE '%машинн%обучен%' OR name ILIKE '%machine%learning%')

ORDER BY 3
'''

In [142]:
df = pd.read_sql_query(query_7_1, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\2740870260.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


schedule avg_slr  cnt
0                              Гибкий график  140500    4
1                           Удаленная работа  190043   17
2                                Полный день  175726   41
3  Итого по всем вакансиям DS с указанной ЗП  177379   62
4                 Итого по всем вакансиям DS          483

Средняя зарплата при удаленной работе для дата-сайентистов выше, чем при полном дне. Хотя такой тип рабочего дня распространен меньше классического полного дня, для вакансий, предполагающих удаленную работу составляют 24% (117 вакансий) от выборки вакансий для DS. В результате переезд в другой город для дата-сайентистов чаще всего не имеет смысла, т.к. можно найти более высокооплачиваемую вакансию, предполагающую удаленную работу.
Помимо этого, стоит обратить внимание, что по большой доле вакансий DS ни в каком виде не указана зарплата, что в будущем может значительно исказить качество модели в части факторов, привязанных к размеру оплаты труда.

2. Рассмотрим географию вакансий, не подразумевающих удаленную работу. Выведем ТОП-10 регионов по количеству вакансий с таким типом рабочего графика.

In [171]:
n = 5
query_7_2 =  f'''
SELECT
a.name,
COUNT(v.id) cnt
FROM vacancies v
JOIN areas a ON v.area_id = a.id
WHERE (v.name ILIKE '%data%scientist%' OR v.name ILIKE '%data%science%' OR v.name ILIKE '%исследователь%данных%' OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
OR v.name ILIKE '%машинн%обучен%' OR v.name ILIKE '%machine%learning%')
AND v.schedule != 'Удаленная работа'
GROUP BY a.name
ORDER BY 2 DESC
LIMIT {n}
'''

In [172]:
df = pd.read_sql_query(query_7_2, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\3686448377.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


name  cnt
0           Москва  173
1  Санкт-Петербург   53
2      Новосибирск   18
3  Нижний Новгород   17
4           Алматы   12

География подразумевающих посещение офиса вакансий для дата-сайентистов еще более сильно сконцентрирована на Москве и Санкт-Петербурге в сравнении с географией всей выборки. Принимая во внимание, что стоимость жизни (living cost) в этих городах заметно выше других городов выборки, и предполагая, что гипотеза из предыдущего запроса о более высокой зарплате при удаленной работе верна, можно предположить повышенный спрос соискателей на удаленную работу.

3. Рассмотрим работодателей, наиболее активно ищущих дата-сайентистов. Выведем ТОП-10 работодателей по количеству релевантных вакансий, а также укажем среднюю зарплату где это возможно (или Н/Д, если зарплата не указывалась).

In [161]:
n = 10
query_7_3 =  f'''
SELECT
e.name,
COUNT(v.id) cnt
FROM vacancies v
JOIN employers e ON v.employer_id = e.id
WHERE (v.name ILIKE '%data%scientist%' OR v.name ILIKE '%data%science%' OR v.name ILIKE '%исследователь%данных%' OR (v.name LIKE '%ML%' AND v.name NOT ILIKE '%HTML%')
OR v.name ILIKE '%машинн%обучен%' OR v.name ILIKE '%machine%learning%')
GROUP BY e.name
ORDER BY 2 DESC
LIMIT {n}
'''

In [162]:
df = pd.read_sql_query(query_7_3, connection)
display(df)

C:\Users\Alexander\AppData\Local\Temp\ipykernel_9912\3686448377.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


name  cnt
0                   СБЕР   38
1        Bell Integrator   25
2         Банк ВТБ (ПАО)   18
3                     VK   15
4  Positive Technologies   11
5                 Яндекс    9
6         EvenBet Gaming    9
7                МегаФон    8
8               Andersen    7
9                   Ozon    6

Среди основных работодателей для дата-сайентистов выступают в основном крупнейшие российские ИТ и телекоммуникационные компании. Обращает на себя тот факт, что крупнейший по количеству вакансий работодатель - Яндекс, в выборке вакансий для DS занимает лишь 6 место. При этом в ТОП-3 попали два банка.

## Общие выводы:
В рамках проекта мы рассмотрели выборку вакансий для IT-специалистов и набор сопутствующих баз данных для изучения и анализа этой информации с целью ее использования в будущем для создания модели машинного обучения, которая будет рекомендовать вакансии клиентам кадрового агентства, претендующим на позицию Data Scientist.

В рассматриваемых базах содержится информация о 49 тыс. вакансиях и 23,5 тыс. компаниях. Как изначально и предполагалось, далеко не по всем из указанным компаниям есть актуальные вакансии. В основном вакансии географически привязаны к Москве и Санкт-Петербурге, что в целом соответствует общей деловой активностью российской экономики.

Сразу бросается в глаза, тот факт, что более чем по половине вакансий отсутствует информация зарплате, что может негативно отразиться на качестве разрабатываемой модели, т.к. зарплатные ожидания один из основных факторов при принятии решения как для соискателя, так и для работодателя. При этом средняя нижняя граница равна 71 тыс. рублей, а средняя верхняя граница - 111 тыс. рублей.

Работодатели преимущественно ищут сотрудников на полную занятость, но часто допускают удаленную работу (вторая по популярности опция). При этом в выборке 83% приходится на вакансии, допускающим опыт менее 3 лет либо вообще отсутствие опыта.

Общая выборка практически не имеет какой-либо концентрации на работодателях. Доля крупнейшего работодателя - Яндекс - 4%, при этом география Яндекс гораздо более равномерна в сравнении с общей выборкой. Обращает на себя внимание отсутствие более чем у трети компаний информации о сфере деятельности. Среди наиболее распространенных - разработка программного обеспечения и системная интеграция, что в целом соответствует общей специфике рассматриваемой выборки.

Учитывая тот факт, что предполагается разработка специализированной модели, необходимо более детально рассмотреть целевую выборку. Из общей выборки 1771 вакансия (или 3,6%) имеет отношение к данным. Это говорит о том, что выявленные закономерности для всей выборки не могут быть экстраполированы на эту подвыборку и необходим детальный анализ таких вакансий. Для дата-сайентистов начального уровня доступно 51 вакансия или 2,9% от всех вакансий, связанных с данными. Среди ключевых навыков для DS указаны знание Python, умение работать с SQL или postgres.

Среди основных работодателей для дата-сайентистов выступают в основном крунейшие российские ИТ и телкоммуникационные компании. Обращает на себя тот факт, что крупнейший по количеству вакансий работодатель - Яндекс, в выборке вакансий для DS занимает лишь 6 место. При этом в ТОП-3 попали два банка. География подразумевающих посещение офиса вакансий для дата-сайентистов еще более сильно сконцентрирована на Москве и Санкт-Петербурге в сравнении с географией всей выборки. 

Для выборки вакансий для DS, даже более чем для общей выборки, выражен фактор отсутствия данных об информации о зарплате. Доля таких вакансий в подвыборке составляет 87% против 51% по всей выборке. Тем не менее, если предположить, что тенденции, выявленные в тех вакансиям, где информация о зарплате указана, релевантны для всей подвыборки DS, то можно сделать следующие вывод:

1. На начальном этапе своей карьеры дата-сайентист может рассчитывать на зарплату около 75 тыс. рублей, после года опыта показатель может увеличиться почти в два раза - до 140 тыс. рублей, тогда как после трех лет опыта зарплаты может приблизиться к 250 тыс. рублей. Это заметно выше аналогичных уровне для всей выборки - 40, 77 и 133 тыс. рублей соответственно.
2. Средняя зарплата при удаленной работе для дата-сайентистов выше, чем при полном дне. Хотя такой тип рабочего дня распространен меньше классического полного дня, для вакансий, предполагающих удаленную работу составляют 24% (117 вакансий) от выборки вакансий для DS. В результате переезд в другой город для дата-сайентистов чаще всего не имеет смысла, т.к. можно найти более высокооплачиваемую вакансию, предполагающую удаленную работу.
3. Принимая во внимание, что (a) стоимость жизни (living cost) в Москве и Санкт-Петербурге заметно выше других городов выборки и (b) зарплата при удаленной работе DS выше других вариантов графика работы, можно предположить повышенный спрос соискателей на удаленную работу.